packages/langium/src/documents/document.ts

```diff
@@ -10,15 +10,57 @@ import { TextDocuments } from 'vscode-languageserver/node';
 import { AstNode } from '../syntax-tree';
 import { LangiumParser, ParseResult } from '../parser/langium-parser';
 import { AstNodeDescription } from '../references/scope';
-import { LangiumServices, LanguageMetaData, Stream, stream } from '..';
 import { URI } from 'vscode-uri';
+import { Mutable } from '../utils/ast-util';
+import { LanguageMetaData } from '../grammar/language-meta-data';
+import { LangiumServices } from '../services';
+import { stream, Stream } from '../utils/stream';
+
+export enum DocumentState {
+    Changed = 0,
+    Parsed = 1,
+    Processed = 2,
+    Indexed = 3,
+    Validated = 4
+}
+
+// = 'changed' | 'parsed' | 'processed' | 'indexed' | 'validated';
 
 export interface LangiumDocument {
     parseResult: ParseResult
     precomputedScopes?: PrecomputedScopes
-    valid: boolean
+    state: DocumentState
     textDocument: TextDocument
     uri: URI
+    reset(): void
+    onReset(listener: () => void): void
+}
+
+export class LangiumDocumentImpl implements LangiumDocument {
+    parseResult: ParseResult<AstNode>;
+    precomputedScopes?: PrecomputedScopes | undefined;
+    state: DocumentState;
+    textDocument: TextDocument;
+    uri: URI;
+
+    private resetListeners: Array<() => void> = [];
+
+    constructor(textDocument: TextDocument, parseResult: ParseResult) {
+        this.parseResult = parseResult;
+        (this.parseResult.value as Mutable<AstNode>).$document = this;
+        this.state = DocumentState.Parsed;
+        this.textDocument = textDocument;
+        this.uri = URI.parse(textDocument.uri);
+    }
+
+    reset(): void {
+        this.state = DocumentState.Parsed;
```

Comment:
When an affected document is reset, it does not need to be reindexed because its exported nodes should not depend on resolving cross-references (see above). So it should be sufficient to reset it to `Indexed`.